algorithms/HPdclassifier/R/predict.hpdegbm.R

```diff
@@ -0,0 +1,218 @@
+####################################################################################
+
+####################################################################################
+#  File hpdegbm.R
+#  Distributed version of gbm 
+
+#  This code is a distributed version of gbm function availabel in R gbm package.
+
+####################################################################################
+## A copy of gbm R package license:
+#
+#  Copyright (C) 1995-2012 The R Core Team
+#
+#  This program is free software; you can redistribute it and/or modify
+#  it under the terms of the GNU General Public License as published by
+#  the Free Software Foundation; either version 2 of the License, or
+#  (at your option) any later version.
+#
+#  This program is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#  GNU General Public License for more details.
+#
+#  A copy of the GNU General Public License is available at
+#  http://www.r-project.org/Licenses/
+####################################################################################
+
+
+####################################################################################
+###                     Data-distributed and model-centralized prediction function of distributed GBM
+####################################################################################
+predict.hpdegbm <- function(object, newdata, appType="binary-classification", type="link", trace = FALSE)
+{
+  # model: assembled GBM model
+  # best.iter: best iteration of sub-models
+  # newdata: testing data in dframe/darray
+  # appType: regression, binary-classification, multi-classification
+  # type="link" or "response"
+
+  ############### data-distributed and model-centralized  prediction #################
+  # suitable for models whose size can fit into the memory
+  # train data: npartition_train/nExecutor_train
+  # test data : npartition_test/nExecutor_test. maybe different from train data
+
+  # check function arguments
+  if(missing(object))
+     stop("'object' is a required argument")
+
+  if(missing(best.iter))
+     stop("'best.iter' is a required argument")
+
+  if(missing(newdata))
+     stop("'newdata' is a required argument")
+
+  if(!is.dframe(newdata) && !is.darray(newdata) && !is.data.frame(newdata) && !is.matrix(newdata))
+     stop("'newdata' must be a dframe or darray or data.frame or matrix")
+ 
+  # extract GBM models and corresponding best iterations (n.trees)
+  model <- object[[1]]
+  best.iter <- object[[2]]
+
+  # prediction for distributed big data 
+  if((is.dframe(newdata)) || (is.darray(newdata))) {
+     npartition_test <- npartitions(newdata)  
+     nExecutor_test  <- npartition_test
+
+     nTest <- nrow(newdata)
+     daPredict <- darray(dim=c(nTest,1), blocks=c(ceiling(nTest/npartition_test),1), sparse=FALSE)
```

Comment:
1) Clearly, your "daPredict" must have exactly the same partitioning schema as "newdata". Your approach for creating "daPredict" will not essentially end up to this desired situation when "newdata" is a flexible darray/dframe. Instead, you should create it as a flexible dobject with specifying only "npartitions".
2) I am really surprised that you always generate a darray for the output of prediction. Don't you support classification?